capstone/src/test/java/com/google/sps/CachedInterestVectorTest.java

```diff
@@ -0,0 +1,148 @@
+        vectorHashMap.put(Tag.COOKING.getTag(), (2.0/magnitude));
+        vectorHashMap.put(Tag.FITNESS.getTag(), (1.0/magnitude));
+
+        EmbeddedEntity embeddedVectorMap = CachedInterestVector.hashMapToEmbeddedEntity(vectorHashMap);
+
+        Assert.assertEquals((double) embeddedVectorMap.getProperty(Tag.EDUCATION.getTag()), (2.0/magnitude), epsilon);
+        Assert.assertEquals((double) embeddedVectorMap.getProperty(Tag.COOKING.getTag()), (2.0/magnitude), epsilon);
+        Assert.assertEquals((double) embeddedVectorMap.getProperty(Tag.FITNESS.getTag()), (1.0/magnitude), epsilon);
+    }
+
+    @Test
+    public void denormalizeVectorHashMapTest() {
+        Double magnitude = 3.0;
+        HashMap<String, Double> vectorHashMap = new HashMap<>();
+        vectorHashMap.put(Tag.EDUCATION.getTag(), (2.0/magnitude));
+        vectorHashMap.put(Tag.COOKING.getTag(), (2.0/magnitude));
+        vectorHashMap.put(Tag.FITNESS.getTag(), (1.0/magnitude));
+
+        CachedInterestVector.denormalizeVectorHashMap(vectorHashMap, magnitude);
+
+        Assert.assertEquals(vectorHashMap.get(Tag.EDUCATION.getTag()), 2.0, epsilon);
+        Assert.assertEquals(vectorHashMap.get(Tag.COOKING.getTag()), 2.0, epsilon);
+        Assert.assertEquals(vectorHashMap.get(Tag.FITNESS.getTag()), 1.0, epsilon);
+    }
+
+    @Test
+    public void addTagToDenormalizedVectorHashMapTest() {
+        HashMap<String, Double> vectorHashMap = new HashMap<>();
+        vectorHashMap.put(Tag.EDUCATION.getTag(), 1.0);
+        vectorHashMap.put(Tag.COOKING.getTag(), 1.0);
+
+        List<Tag> tagsToAdd = new ArrayList<>(); 
+        tagsToAdd.add(Tag.EDUCATION);
+        tagsToAdd.add(Tag.COOKING);
+        tagsToAdd.add(Tag.FITNESS);
+
+        CachedInterestVector.addTagToDenormalizedVectorHashMap(vectorHashMap, tagsToAdd);
+
+        Assert.assertEquals(vectorHashMap.get(Tag.EDUCATION.getTag()), 2.0, epsilon);
+        Assert.assertEquals(vectorHashMap.get(Tag.COOKING.getTag()), 2.0, epsilon);
+        Assert.assertEquals(vectorHashMap.get(Tag.FITNESS.getTag()), 1.0, epsilon);
+    }
+
+    @Test
+    public void removeTagFromDenormalizedVectorHashMapTest() {
+        HashMap<String, Double> vectorHashMap = new HashMap<>();
+        vectorHashMap.put(Tag.EDUCATION.getTag(), 2.0);
+        vectorHashMap.put(Tag.COOKING.getTag(), 2.0);
+        vectorHashMap.put(Tag.FITNESS.getTag(), 1.0);
+
+        List<Tag> tagsToRemove = new ArrayList<>(); 
+        tagsToRemove.add(Tag.EDUCATION);
+        tagsToRemove.add(Tag.COOKING);
+        tagsToRemove.add(Tag.FITNESS);
+
+        CachedInterestVector.removeTagFromDenormalizedVectorHashMap(vectorHashMap, tagsToRemove);
+
+        Assert.assertEquals(vectorHashMap.get(Tag.EDUCATION.getTag()), 1.0, epsilon);
+        Assert.assertEquals(vectorHashMap.get(Tag.COOKING.getTag()), 1.0, epsilon);
+        Assert.assertNull(vectorHashMap.get(Tag.FITNESS.getTag()));
+    }
```

Comment:
Test (this or another) should cover what happens when the tag was already not present, but is asked to be removed anyway